File: Evalution.Test/AstBuilderTest.fs
namespace Evalution.Test
open Evalution
open NUnit.Framework

[<TestFixture>]
type AstBuilderTest() =
    
    [<Test>]
    member x.TestAddition ()=
        let result = AstBuilder.build "1+2"
        
        let expectedResult =
            Ast.BinaryExpression(
                Ast.LiteralExpression(Ast.Int32Literal(1)),
                Ast.Add,
                Ast.LiteralExpression(Ast.Int32Literal(2)))

        Assert.AreEqual(expectedResult, result)

    [<Test>]
    member x.TestSubtract ()=
        let result = AstBuilder.build "1-2"
        
        let expectedResult =
            Ast.BinaryExpression(
                Ast.LiteralExpression(Ast.Int32Literal(1)),
                Ast.Subtract,
                Ast.LiteralExpression(Ast.Int32Literal(2)))

        Assert.AreEqual(expectedResult, result)

    [<Test>]
    member x.TestMultiply ()=
        let result = AstBuilder.build "1*2"
        
        let expectedResult =
            Ast.BinaryExpression(
                Ast.LiteralExpression(Ast.Int32Literal(1)),
                Ast.Multiply,
                Ast.LiteralExpression(Ast.Int32Literal(2)))

        Assert.AreEqual(expectedResult, result)

    [<Test>]
    member x.TestDivide ()=
        let result = AstBuilder.build "1/2"
        
        let expectedResult =
            Ast.BinaryExpression(
                Ast.LiteralExpression(Ast.Int32Literal(1)),
                Ast.Divide,
                Ast.LiteralExpression(Ast.Int32Literal(2)))

        Assert.AreEqual(expectedResult, result)

    [<Test>]
    member x.TestInt32Literal ()=
        let result = AstBuilder.build "1"
        
        let expectedResult =
            Ast.LiteralExpression(Ast.Int32Literal(1))

        Assert.AreEqual(expectedResult, result)

    [<Test>]
    member x.TestDoubleLiteral ()=
        let result = AstBuilder.build "1.5"
        
        let expectedResult =
            Ast.LiteralExpression(Ast.DoubleLiteral(1.5))

        Assert.AreEqual(expectedResult, result)

    [<Test>]
    member x.TestBinaryExpressionInt32DivideDouble ()=
        let result = AstBuilder.build "3/2"
        
        let expectedResult =
            Ast.BinaryExpression(
                Ast.LiteralExpression(Ast.Int32Literal(3)),
                Ast.Divide,
                Ast.LiteralExpression(Ast.Int32Literal(2)))

        Assert.AreEqual(expectedResult, result)

    [<Test>]
    member x.TestSpaces ()=
        let result = AstBuilder.build "  1.5 +	2 "
        
        let expectedResult =
            Ast.BinaryExpression(
                Ast.LiteralExpression(Ast.DoubleLiteral(1.5)),
                Ast.Add,
                Ast.LiteralExpression(Ast.Int32Literal(2)))

        Assert.AreEqual(expectedResult, result)

    [<Test>]
    member x.TestParentheses()=
        let result = AstBuilder.build "( 1.5 + 1 ) * 2"
        
        let expectedResult =
            Ast.BinaryExpression(
                Ast.BinaryExpression(
                    Ast.LiteralExpression(Ast.DoubleLiteral(1.5)),
                    Ast.Add,
                    Ast.LiteralExpression(Ast.Int32Literal(1))),
                Ast.Multiply,
                Ast.LiteralExpression(Ast.Int32Literal(2)))

        Assert.AreEqual(expectedResult, result)

    [<Test>]
    member x.TestMethodCallExpression ()=
        let result = AstBuilder.build "method(1.0,2+3)"

        let expectedResult =
            Ast.MultiCallExpression (
                Ast.CurrentContextMethodCall(
                    "method",
                    [
                        Ast.LiteralExpression(
                            Ast.DoubleLiteral(1.0)
                        );
                        Ast.BinaryExpression(
                            Ast.LiteralExpression(Ast.Int32Literal(2)),
                            Ast.Add,
                            Ast.LiteralExpression(Ast.Int32Literal(3))
                        )
                    ]
                )
            )

        Assert.AreEqual(expectedResult, result)

    [<Test>]
    member x.TestMethodChainCallExpression1 ()=
        let result = AstBuilder.build "Method1(1).Method2(2)"

        let expectedResult =
            Ast.MultiCallExpression(
                Ast.ObjectContextMethodCall(
                    Ast.CurrentContextMethodCall(
                        "Method1", [Ast.LiteralExpression(Ast.Int32Literal(1))]
                    ), "Method2", [Ast.LiteralExpression(Ast.Int32Literal(2))]
                )
            )

        Assert.AreEqual(expectedResult, result)

    [<Test>]
    member x.TestMethodChainCallExpression2 ()=
        let result = AstBuilder.build "Method1(1).Prop1.Method2(2)"

        let expectedResult =
            Ast.MultiCallExpression(
                Ast.ObjectContextMethodCall(
                    Ast.ObjectContextPropertyCall(
                        Ast.CurrentContextMethodCall(
                            "Method1", [Ast.LiteralExpression(Ast.Int32Literal(1))]
                        ),
                        "Prop1" ),
                    "Method2", [Ast.LiteralExpression(Ast.Int32Literal(2))]
                )
            )

        Assert.AreEqual(expectedResult, result)

    [<Test>]
    member x.TestMethodCallExpression_NoArguments ()=
        let result = AstBuilder.build "method()"

        let expectedResult =
            Ast.MultiCallExpression (
                Ast.CurrentContextMethodCall(
                    "method",
                    []
                )
            )

        Assert.AreEqual(expectedResult, result)

    [<Test>]
    member x.TestMultiCallExpression_Single ()=
        let result = AstBuilder.build "var1"
        
        let expectedResult =
            Ast.MultiCallExpression(
                Ast.CurrentContextPropertyCall("var1"))

        Assert.AreEqual(expectedResult, result)

    [<Test>]
    member x.TestMultiCallExpression_Multiple ()=
        let result = AstBuilder.build "var1.var2.var3"
        
        let expectedResult =
            Ast.MultiCallExpression(
                Ast.ObjectContextPropertyCall(
                    Ast.ObjectContextPropertyCall(
                        Ast.CurrentContextPropertyCall("var1")
                        , "var2"
                    ), "var3"
                )
            )

        Assert.AreEqual(expectedResult, result)

    [<Test>]
    member x.TestArrayExpression ()=
        let result = AstBuilder.build "array[12].Item"
        
        let expectedResult =
            Ast.MultiCallExpression(
                Ast.ObjectContextPropertyCall(
                    Ast.ArrayElementCall(
                        Ast.CurrentContextPropertyCall("array"),
                        Ast.LiteralExpression(Ast.Int32Literal(12))
                    ),
                    "Item"
                )
            )

        Assert.AreEqual(expectedResult, result)

    [<Test>]
    member x.TestArrayExpression_Nested ()=
        let result = AstBuilder.build "array[12][3]"
        
        let expectedResult =
            Ast.MultiCallExpression(
                Ast.ArrayElementCall(
                    Ast.ArrayElementCall(
                        Ast.CurrentContextPropertyCall("array"),
                        Ast.LiteralExpression(Ast.Int32Literal(12))
                    ),
                    Ast.LiteralExpression(Ast.Int32Literal(3))
                )
            )

        Assert.AreEqual(expectedResult, result)

    [<Test>]
    member x.TestUnaryOperatorNegate()=
        let result = AstBuilder.build "-2 + 1"
        
        let expectedResult =
            Ast.BinaryExpression(
                Ast.UnaryExpression(Ast.Negate, Ast.LiteralExpression(Ast.Int32Literal(2))),
                Ast.Add,
                Ast.LiteralExpression(Ast.Int32Literal(1)))

        Assert.AreEqual(expectedResult, result)

    [<Test>]
    member x.TestUnaryOperatorAndParenthesis()=
        let result = AstBuilder.build "-(2+2)+2"
        
        let expectedResult =
            Ast.BinaryExpression(
                Ast.UnaryExpression(
                    Ast.Negate,
                    Ast.BinaryExpression(
                        Ast.LiteralExpression(Ast.Int32Literal(2)),
                        Ast.Add,
                        Ast.LiteralExpression(Ast.Int32Literal(2)))),
                Ast.Add,
                Ast.LiteralExpression(Ast.Int32Literal(2)))

        Assert.AreEqual(expectedResult, result)

    [<Test>]
    member x.TestUnaryOperatorIdentity()=
        let result = AstBuilder.build "+2 + 1"
        
        let expectedResult =
            Ast.BinaryExpression(
                Ast.UnaryExpression(Ast.Identity, Ast.LiteralExpression(Ast.Int32Literal(2))),
                Ast.Add,
                Ast.LiteralExpression(Ast.Int32Literal(1)))

        Assert.AreEqual(expectedResult, result)

    [<Test>]
    member x.TestUnaryOperatorLogicalNegate()=
        // TODO: this is actually is not correct, because LogicalNegate can be applied only to bool value
        let result = AstBuilder.build "!2 + 1"
        
        let expectedResult =
            Ast.BinaryExpression(
                Ast.UnaryExpression(Ast.LogicalNegate, Ast.LiteralExpression(Ast.Int32Literal(2))),
                Ast.Add,
                Ast.LiteralExpression(Ast.Int32Literal(1)))

        Assert.AreEqual(expectedResult, result)

File: src/Evalution.CSharpTest/AstBuilderTest.cs
namespace Evalution.CSharpTest
{
    using System.Collections.Generic;
    using EvalutionCS.Ast;
    using NUnit.Framework;

    [TestFixture]
    public class AstBuilderTest
    {
        [Test]
        public void TestAddition()
        {
            var result = AstBuilder.Build("1+2");

            var expectedResult =
                new BinaryExpression(
                    new LiteralExpression(new Int32Literal(1)),
                    BinaryOperator.Add,
                    new LiteralExpression(new Int32Literal(2)));

            Assert.AreEqual(expectedResult, result);;
        }

        [Test]
        public void TestSubtract()
        {
            var result = AstBuilder.Build("1-2");

            var expectedResult =
                new BinaryExpression(
                    new LiteralExpression(new Int32Literal(1)),
                    BinaryOperator.Subtract,
                    new LiteralExpression(new Int32Literal(2)));

            Assert.AreEqual(expectedResult, result);
        }

        [Test]
        public void TestMultiply()
        {
            var result = AstBuilder.Build("1*2");


            var expectedResult =
                new BinaryExpression(
                    new LiteralExpression(new Int32Literal(1)),
                    BinaryOperator.Multiply,
                    new LiteralExpression(new Int32Literal(2)));

            Assert.AreEqual(expectedResult, result);
        }

        [Test]
        public void TestDivide()
        {
            var result = AstBuilder.Build("1/2");


            var expectedResult =
                new BinaryExpression(
                    new LiteralExpression(new Int32Literal(1)),
                    BinaryOperator.Divide,
                    new LiteralExpression(new Int32Literal(2)));

            Assert.AreEqual(expectedResult, result);
        }

        [Test]
        public void TestInt32Literal()
        {
            var result = AstBuilder.Build("1");


            var expectedResult =
                new LiteralExpression(new Int32Literal(1));

            Assert.AreEqual(expectedResult, result);
        }

        [Test]
        public void TestDoubleLiteral()
        {
            var result = AstBuilder.Build("1.5");


            var expectedResult =
                new LiteralExpression(new DoubleLiteral(1.5));

            Assert.AreEqual(expectedResult, result);
        }

        [Test]
        public void TestBinaryExpressionInt32DivideDouble()
        {
            var result = AstBuilder.Build("3/2");


            var expectedResult =
                new BinaryExpression(
                    new LiteralExpression(new Int32Literal(3)),
                    BinaryOperator.Divide,
                    new LiteralExpression(new Int32Literal(2)));

            Assert.AreEqual(expectedResult, result);
        }

        [Test]
        public void TestSpaces()
        {
            var result = AstBuilder.Build("  1.5 +	2 ");


            var expectedResult =
                new BinaryExpression(
                    new LiteralExpression(new DoubleLiteral(1.5)),
                    BinaryOperator.Add,
                    new LiteralExpression(new Int32Literal(2)));

            Assert.AreEqual(expectedResult, result);
        }

        [Test]
        public void TestParentheses()
        {
            var result = AstBuilder.Build("( 1.5 + 1 ) * 2");

            var expectedResult =
                new BinaryExpression(
                    new BinaryExpression(
                        new LiteralExpression(new DoubleLiteral(1.5)),
                        BinaryOperator.Add,
                        new LiteralExpression(new Int32Literal(1))),
                    BinaryOperator.Multiply,
                    new LiteralExpression(new Int32Literal(2)));

            Assert.AreEqual(expectedResult, result);
        }

        [Test]
        public void TestMethodCallExpression(){
            var result = AstBuilder.Build("method(1.0,2+3)");

            var expectedResult =
                new MultiCallExpression(
                    new CurrentContextMethodCall(
                        "method",
                        new List<Expression>()
                        {
                            new LiteralExpression(
                                new DoubleLiteral(1.0)
                                ),
                            new BinaryExpression(
                                new LiteralExpression(new Int32Literal(2)),
                                BinaryOperator.Add,
                                new LiteralExpression(new Int32Literal(3))
                                )
                        }
                        )
                    );

            Assert.AreEqual(expectedResult, result);
        }
        [Test]
        public void TestMethodChainCallExpression1(){
            var result = AstBuilder.Build("Method1(1).Method2(2)"); 

            var expectedResult =
                new MultiCallExpression(
                    new ObjectContextMethodCall(
                        new CurrentContextMethodCall(
                            "Method1", new List<Expression>(){new LiteralExpression(new Int32Literal(1))}
                        ), "Method2", new List<Expression>{new LiteralExpression(new Int32Literal(2))}
                    )
                );

            Assert.AreEqual(expectedResult, result);
        }
        [Test]
        public void TestMethodChainCallExpression2(){
            var result = AstBuilder.Build("Method1(1).Prop1.Method2(2)"); 

            var expectedResult =
                new MultiCallExpression(
                    new ObjectContextMethodCall(
                        new ObjectContextPropertyCall(
                            new CurrentContextMethodCall(
                                "Method1", new List<Expression>{new LiteralExpression(new Int32Literal(1))}
                            ),
                            "Prop1" ),
                        "Method2", new List<Expression> { new LiteralExpression(new Int32Literal(2)) }
                    )
                );

            Assert.AreEqual(expectedResult, result);
        }
        [Test]
        public void TestMethodCallExpression_NoArguments(){
            var result = AstBuilder.Build("method()");

            var expectedResult =
                new MultiCallExpression(
                    new CurrentContextMethodCall(
                        "method",
                        new List<Expression>()
                        )
                    );

            Assert.AreEqual(expectedResult, result);
        }

        [Test]
        public void TestMultiCallExpression_Single()
        {
            var result = AstBuilder.Build("var1");


            var expectedResult =
                new MultiCallExpression(
                    new CurrentContextPropertyCall("var1"));

            Assert.AreEqual(expectedResult, result);
        }

        [Test]
        public void TestMultiCallExpression_Multiple()
        {
            var result = AstBuilder.Build("var1.var2.var3");


            var expectedResult =
                new MultiCallExpression(
                    new ObjectContextPropertyCall(
                        new ObjectContextPropertyCall(
                            new CurrentContextPropertyCall("var1")
                            , "var2"
                            ), "var3"
                        )
                    );

            Assert.AreEqual(expectedResult, result);
        }

        [Test]
        public void TestArrayExpression()
        {
            var result = AstBuilder.Build("array[12].Item");


            var expectedResult =
                new MultiCallExpression(
                    new ObjectContextPropertyCall(
                        new ArrayElementCall(
                            new CurrentContextPropertyCall("array"),
                            new LiteralExpression(new Int32Literal(12))
                            ),
                        "Item"
                        )
                    );

            Assert.AreEqual(expectedResult, result);
        }

        [Test]
        public void TestArrayExpression_Nested()
        {
            var result = AstBuilder.Build("array[12][3]");


            var expectedResult =
                new MultiCallExpression(
                    new ArrayElementCall(
                        new ArrayElementCall(
                            new CurrentContextPropertyCall("array"),
                            new LiteralExpression(new Int32Literal(12))
                            ),
                        new LiteralExpression(new Int32Literal(3))
                        )
                    );

            Assert.AreEqual(expectedResult, result);
        }

        [Test]
        public void TestUnaryOperatorNegate()
        {
            var result = AstBuilder.Build("-2 + 1");

            var expectedResult =
                new BinaryExpression(
                    new UnaryExpression(UnaryOperator.Negate, new LiteralExpression(new Int32Literal(2))),
                    BinaryOperator.Add,
                    new LiteralExpression(new Int32Literal(1)));

            Assert.AreEqual(expectedResult, result);
        }

        [Test]
        public void TestUnaryOperatorAndParenthesis()
        {
            var result = AstBuilder.Build("-(2+2)+2");


            var expectedResult =
                new BinaryExpression(
                    new UnaryExpression(
                        UnaryOperator.Negate,
                        new BinaryExpression(
                            new LiteralExpression(new Int32Literal(2)),
                            BinaryOperator.Add,
                            new LiteralExpression(new Int32Literal(2)))),
                    BinaryOperator.Add,
                    new LiteralExpression(new Int32Literal(2)));

            Assert.AreEqual(expectedResult, result);
        }

        [Test]
        public void TestUnaryOperatorIdentity()
        {
            var result = AstBuilder.Build("+2 + 1");


            var expectedResult =
                new BinaryExpression(
                    new UnaryExpression(UnaryOperator.Identity, new LiteralExpression(new Int32Literal(2))),
                    BinaryOperator.Add,
                    new LiteralExpression(new Int32Literal(1)));

            Assert.AreEqual(expectedResult, result);
        }

        [Test]
        public void TestUnaryOperatorLogicalNegate()
        {
            // TODO: this is actually is not correct, because LogicalNegate can be applied only to bool value
            var result = AstBuilder.Build("!2 + 1");


            var expectedResult =
                new BinaryExpression(
                    new UnaryExpression(UnaryOperator.LogicalNegate, new LiteralExpression(new Int32Literal(2))),
                    BinaryOperator.Add,
                    new LiteralExpression(new Int32Literal(1)));

            Assert.AreEqual(expectedResult, result);
        }
    }
}